utils/sites.js

```diff
@@ -24,60 +23,121 @@ const submitParticipantsData = async (req, res, site) => {
         if(!authorized){
             return res.status(401).json(getResponseJSON('Authorization failed!', 401));
         }
+
         siteCode = authorized.siteCode;
     }
 
-    if(req.body.data === undefined) return res.status(400).json(getResponseJSON('data is undefined in request body.', 400));
-    if(req.body.data.length === undefined || req.body.data.length < 1) return res.status(400).json(getResponseJSON('data array doesn\'t have any element.', 400));
+    if(req.body.data === undefined) return res.status(400).json(getResponseJSON('Bad request. Data is not defined in request body.', 400));
+    if(!Array.isArray(req.body.data)) return res.status(400).json(getResponseJSON('Bad request. Data must be an array.', 400));
+    if(req.body.data.length === undefined || req.body.data.length < 1) return res.status(400).json(getResponseJSON('Bad request. Data array does not have any elements.', 400));
+    if(req.body.data.length > 499) return res.status(400).json(getResponseJSON('Bad request. Data contains more than acceptable limit of 500 records.', 400));
 
-    if(req.body.data.length > 499) return res.status(400).json(getResponseJSON('More than acceptable limit of 500 records.', 400));
+    console.log(req.body.data);
+    
+    const dataArray = req.body.data;
 
-    const data = req.body.data;
-    console.log(`${JSON.stringify(data)}`);
+    let responseArray = [];
     let error = false;
-    let errorMsgs = [];
-    for(let obj of data){
-        if(obj.token){
-            const participantToken = obj.token;
-            const { getParticipantData } = require('./firestore');
-            const record = await getParticipantData(participantToken, siteCode);
-            if(record){
-                const docID = record.id;
-                delete obj.token;
-                let newStateElements = {}
-                for(let key in obj) {
-                    if(record.data.state[key] === undefined){
-                        newStateElements[`state.${key}`] = obj[key];
+
+    //let errorMsgs = [];
+
+    for(let dataObj of dataArray){
+        if(dataObj.token === undefined) {
+            error = true;
+            responseArray.push({'Invalid Request': {'Token': 'UNDEFINED', 'Errors': 'Token not defined in data object.'}});
+            continue;
+        }
+        
+        const participantToken = dataObj.token;
+        const { getParticipantData } = require('./firestore');
+        const record = await getParticipantData(participantToken, siteCode);
```

Comment:
Should we get data of participants in one query?

Reply:
the way you build / run queries in Firestore I'm not sure it would make a difference.. Firestore doesn't support OR conditionality, and I believe if you used "IN" it would be the same as running an individual query for each one.. I may be mistaken though.